src/Features/LanguageServer/Protocol/Features/EditAndContinue/EditAndContinueDiagnosticSource.cs

```diff
@@ -2,19 +2,77 @@
 
+using System;
 using System.Collections.Immutable;
 using System.Diagnostics;
+using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.Diagnostics;
 using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
+using Microsoft.CodeAnalysis.PooledObjects;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.EditAndContinue;
 
 internal sealed class EditAndContinueDiagnosticSource
 {
-    public static async Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(Document designTimeDocument, CancellationToken cancellationToken)
+    private sealed class ClosedDocumentSource(TextDocument document, ImmutableArray<DiagnosticData> diagnostics) : AbstractWorkspaceDocumentDiagnosticSource(document)
+    {
+        public override bool IsLiveSource()
+            => true;
+
+        public override Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(IDiagnosticAnalyzerService diagnosticAnalyzerService, RequestContext context, CancellationToken cancellationToken)
+            => Task.FromResult(diagnostics);
+    }
+
+    private sealed class ProjectSource(Project project, ImmutableArray<DiagnosticData> diagnostics) : AbstractProjectDiagnosticSource(project)
+    {
+        public override bool IsLiveSource()
+            => true;
+
+        public override Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(IDiagnosticAnalyzerService diagnosticAnalyzerService, RequestContext context, CancellationToken cancellationToken)
+            => Task.FromResult(diagnostics);
+    }
+
+    public static async ValueTask AddWorkspaceDiagnosticSourcesAsync(ArrayBuilder<IDiagnosticSource> sources, Solution solution, Func<Document, bool> isDocumentOpen, CancellationToken cancellationToken)
+    {
+        // avoid creating and synchronizing compile-time solution if Hot Reload/EnC session is not active
+        if (solution.Services.GetRequiredService<IEditAndContinueWorkspaceService>().SessionTracker is not { IsSessionActive: true } sessionStateTracker)
+        {
+            return;
+        }
+
+        var applyDiagnostics = sessionStateTracker.ApplyChangesDiagnostics;
+
+        var dataByDocument = from data in applyDiagnostics
+                             where data.DocumentId != null
+                             group data by data.DocumentId into documentData
+                             select documentData;
+
+        // diagnostics associated with closed documents:
+        foreach (var (documentId, diagnostics) in dataByDocument)
+        {
+            var document = await solution.GetDocumentAsync(documentId, includeSourceGenerated: true, cancellationToken).ConfigureAwait(false);
+            if (document != null && !isDocumentOpen(document))
+            {
+                sources.Add(new ClosedDocumentSource(document, diagnostics.ToImmutableArray()));
```

Comment:
question - so it looks like here we're only adding sources for documents that have EnC diagnostics (seems reasonable, so we don't iterate every file in the sln).  When these diagnostics go away though - how does the report with empty diagnostics for the document get created to clear the old EnC diagnostics?

Reply:
I guess it does not :(. Any suggestions? On diagnostic refresh, would LSP client ask us for new diagnostics for every document we reported diagnostics on?